### PR TITLE
kubeadm: using struct option rather than a long list of parameters

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubeadm/issues/2941

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thanks for the pr.

i think https://github.com/kubernetes/kubeadm/issues/2941 gives reset configuration as an example but the change must be done for init and join configuration too.

**Comment by carlory** [comment]
@neolit123 updated. please review again. thanks.

**Review by neolit123 — COMMENTED** [review]
maybe just use the same struct for all config types configutil.LoadOrDefaultOptions
?

**Comment by carlory** [comment]
I'm not sure if those options hold the same fields. 

cc @pacoxu, what do you think?

**Comment by neolit123** [comment]
> I'm not sure if those struct hold the same fields.
> 
> cc @pacoxu, what do you think?

a struct can hold all the options shared between functions of common usage.

**Comment by carlory** [comment]
@neolit123 updated.

**Review by neolit123 — COMMENTED** [review]
SGTM, thanks

/approve

**Comment by neolit123** [comment]
also adding:
/cc @SataQiu  @chendave 
do we need to migrate any other function arguments into new option struct fields and keep the linked issue open?
e.g. the "upgrade" code has a lot of these.

**Comment by SataQiu** [comment]
> do we need to migrate any other function arguments into new option struct fields and keep the linked issue open?

I'm in favor of this refactor. But it's not very urgent. We can keep the issue open as a long-term task.

**Comment by chendave** [comment]
/retitle: kubeadm: using struct option rather than a long list of parameters

**Comment by chendave** [comment]
/retitle kubeadm: using struct option rather than a long list of parameters

**Review by chendave — COMMENTED** [review]
The code change itself LGTM, two questions here:

- Does the change resolve the original issue: "too deepin code path"
- Does the change simplify the code, particularly, reduce the length of the parameters list?

**Comment by neolit123** [comment]
> Does the change resolve the original issue: "too deepin code path"

@carlory seems to have already changed the "fixes" to "part of". we can track what was done and at some point stop the refactors.

what other target functions we can do next?

> Does the change simplify the code, particularly, reduce the length of the parameters list?

for the LoadOrDefault* functions maybe not too much, but in other places this pattern can help a lot.

**Comment by pacoxu** [comment]
/priority backlog
/triage accepted
LGTM
/assign @SataQiu 
for final approve

File: cmd/kubeadm/app/util/config/common.go [hunk 1]
```diff
@@ -44,6 +44,11 @@ import (
 	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 )
 
+type LoadOrDefaultConfigurationOptions struct {
```
Comment (SataQiu) [inline cmd/kubeadm/app/util/config/common.go:47]:
```suggestion
// LoadOrDefaultConfigurationOptions holds the common LoadOrDefaultConfiguration options.
type LoadOrDefaultConfigurationOptions struct {
```

Comment (SataQiu) [reply cmd/kubeadm/app/util/config/common.go:47]: Missing a comment for the exported type.

File: cmd/kubeadm/app/util/config/common.go [hunk 2]
```diff
@@ -44,6 +44,11 @@ import (
...
+	AllowExperimental bool
```
Comment (SataQiu) [inline cmd/kubeadm/app/util/config/common.go:48]: also the exported fields

Comment (carlory) [reply cmd/kubeadm/app/util/config/common.go:50]: @SataQiu  added.

File: cmd/kubeadm/app/util/config/common.go [hunk 3]
```diff
@@ -44,6 +44,14 @@ import (
 	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 )
 
+// LoadOrDefaultConfigurationOptions holds the common LoadOrDefaultConfiguration options.
+type LoadOrDefaultConfigurationOptions struct {
+	// Allow migration to experimental, unreleased APIs
```
Comment (SataQiu) [inline cmd/kubeadm/app/util/config/common.go:49]: The comment should start with the field name, it's a good practice, such as:
```suggestion
	// AllowExperimental indicates whether the experimental / work in progress APIs can be used.
```

File: cmd/kubeadm/app/util/config/common.go [hunk 4]
```diff
@@ -44,6 +44,14 @@ import (
...
+	AllowExperimental bool
+	// Skip CRI socket detection when the Node API object doesn't contain a CRI socket information
```
Comment (SataQiu) [inline cmd/kubeadm/app/util/config/common.go:51]:
```suggestion
	// SkipCRIDetect indicates whether to skip the CRI socket detection when no CRI socket is provided.
```

**Review by SataQiu — COMMENTED** [review]
thanks for the update
/lgtm
/approve

**Pull request merged**